library/modules/EventManager.cpp

```diff
@@ -703,63 +898,160 @@ static void manageBuildingEvent(color_ostream& out) {
+        // continue if we found the id in world->buildings.all
+        if (index != -1) {
+            continue;
+        }
+        // pretty sure we'd invalidate our loop if we added to buildings here, so we just save the id in an intermediary for now
+        destroyedBuildings.emplace(tick, id);
+        createdBuildings.erase(id);
+    }
+    multimap<Plugin*, EventHandler> copy(handlers[EventType::DESTROYED_BUILDING].begin(), handlers[EventType::DESTROYED_BUILDING].end());
+    // iterate event handler callbacks
+    for (auto &key_value : copy) {
+        auto &handler = key_value.second;
+        auto last_tick = eventLastTick[handler];
+        // enforce handler's callback frequency
+        if (tick - last_tick >= handler.freq) {
+            eventLastTick[handler] = tick;
+            // send the handler all the new & destroyed buildings since it last fired
+            auto jter = destroyedBuildings.upper_bound(last_tick);
+            for (; jter != destroyedBuildings.end(); ++jter) {
+                handler.eventHandler(out, (void*) intptr_t(jter->second));
+            }
+        }
+    }
+}
+
+
+
 static void manageConstructionEvent(color_ostream& out) {
     if (!df::global::world)
         return;
+    // todo: finish the job
     //unordered_set<df::construction*> constructionsNow(df::global::world->constructions.begin(), df::global::world->constructions.end());
+    for (auto constru_iter = constructions.begin(); constru_iter != constructions.end();) {
+        df::construction &construction = (*constru_iter).second;
+        if (df::construction::find(construction.pos) != NULL) {
+            ++constru_iter;
+            continue;
+        }
+        //construction removed
+        //out.print("Removed construction (%d,%d,%d)\n", construction.pos.x,construction.pos.y,construction.pos.z);
+        handler.eventHandler(out, (void*) construction);
+        constru_iter = constructions.erase(constru_iter);
+    }
 
     multimap<Plugin*,EventHandler> copy(handlers[EventType::CONSTRUCTION].begin(), handlers[EventType::CONSTRUCTION].end());
     int32_t tick = df::global::world->frame_counter;
-    for (auto &iter : copy) {
-        auto &handler = iter.second;
-        if (tick - eventLastTick[handler.eventHandler] >= handler.freq) {
-            eventLastTick[handler.eventHandler] = tick;
-            for (auto constru_iter = constructions.begin(); constru_iter != constructions.end();) {
-                df::construction &construction = (*constru_iter).second;
-                if (df::construction::find(construction.pos) != NULL) {
-                    ++constru_iter;
-                    continue;
-                }
-                //construction removed
-                //out.print("Removed construction (%d,%d,%d)\n", construction.pos.x,construction.pos.y,construction.pos.z);
-                handler.eventHandler(out, (void*) &construction);
-                constru_iter = constructions.erase(constru_iter);
-            }
```

Comment:
@expwnent Hoping you can tell me how this is able to work. If the construction object isn't available in DF anymore and we haven't cloned it like we do with `df::job*` in the job events.. then how is it still valid? Can it be expected to stay valid long term?

Reply:
The construction being sent to the handler is the copy we have in our local map (`constructions`)

Reply:
It might be worth renaming our local caches to make this more clear. E.g. rename `constructions` to `constructions_cache` or something